pkg/resources/kafka/headlessService_test.go

```diff
@@ -0,0 +1,242 @@
+// Copyright © 2023 Cisco Systems, Inc. and/or its affiliates
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kafka
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/banzaicloud/koperator/api/v1beta1"
+)
+
+func TestCheckIfNodePortSvcNeeded(t *testing.T) {
+	testReconciler := Reconciler{}
+
+	testCases := []struct {
+		testName            string
+		kafkaCluster        *v1beta1.KafkaCluster
+		isNodePortSvcNeeded bool
+	}{
+		{
+			testName: "KafkaCluster with all external listeners using NodePort",
+			kafkaCluster: &v1beta1.KafkaCluster{
+				Spec: v1beta1.KafkaClusterSpec{
+					ListenersConfig: v1beta1.ListenersConfig{
+						ExternalListeners: []v1beta1.ExternalListenerConfig{
+							{
+								AccessMethod: corev1.ServiceTypeNodePort,
+							},
+							{
+								AccessMethod: corev1.ServiceTypeNodePort,
+							},
+						},
+					},
+				},
+			},
+			isNodePortSvcNeeded: true,
+		},
+		{
+			testName: "KafkaCluster with all external listeners using non-NodePort",
+			kafkaCluster: &v1beta1.KafkaCluster{
+				Spec: v1beta1.KafkaClusterSpec{
+					ListenersConfig: v1beta1.ListenersConfig{
+						ExternalListeners: []v1beta1.ExternalListenerConfig{
+							{
+								AccessMethod: corev1.ServiceTypeLoadBalancer,
+							},
+							{
+								AccessMethod: corev1.ServiceTypeClusterIP,
+							},
+						},
+					},
+				},
+			},
+			isNodePortSvcNeeded: false,
+		},
+		{
+			testName: "KafkaCluster with external listeners of mixed usages of NodePort and non-NodePort",
+			kafkaCluster: &v1beta1.KafkaCluster{
+				Spec: v1beta1.KafkaClusterSpec{
+					ListenersConfig: v1beta1.ListenersConfig{
+						ExternalListeners: []v1beta1.ExternalListenerConfig{
+							{
+								AccessMethod: corev1.ServiceTypeLoadBalancer,
+							},
+							{
+								AccessMethod: corev1.ServiceTypeNodePort,
+							},
+						},
+					},
+				},
+			},
+			isNodePortSvcNeeded: true,
+		},
+	}
+
+	for _, test := range testCases {
+		testReconciler.KafkaCluster = test.kafkaCluster
+		got := testReconciler.checkIfNodePortSvcNeeded()
+		t.Run(test.testName, func(t *testing.T) {
+			if got != test.isNodePortSvcNeeded {
+				t.Errorf("Expected: %v  Got: %v", test.isNodePortSvcNeeded, got)
+			}
+		})
```

Comment:
**Recommended**: We **SHOULD** consider using [`testify/require`](https://pkg.go.dev/github.com/stretchr/testify/require) for improved readability of the test check. I'm going to combine this with my previous comments in the code suggestion.

```suggestion
		t.Run(test.testName, func(t *testing.T) {
			actualHasAnyNodePortAccessMethodExternalListener := testReconciler.checkIfNodePortSvcNeeded()
			require.Equals(t, test.expectedHasAnyNodePortAccessMethodExternalListener, actualHasAnyNodePortAccessMethodExternalListener)
		})
```

Testify also provides a similar error message you manually constructed using the `Equal(...)` function:
```
# https://go.dev/play/p/FqfUImA461T
=== RUN   TestLastIndex/single_item_slice,_last_index_is_first_and_last
    prog_test.go:39: 
        	Error Trace:	/prog_test.go:39
        	Error:      	Not equal: 
        	            	expected: 0
        	            	actual  : -1
        	Test:       	TestLastIndex/single_item_slice,_last_index_is_first_and_last
```

Reply:
I forgot to use `require.Equal` here as I updated the second function to use it rather than `t.Errorf`